### PR TITLE
[Feature] Box oauth refactor

### PR DESCRIPTION
**Pull request by mfraezz**
Purpose
=======
Update the Box addon to use the new OAuth system for additional security.

Changes
=======
Files in /website/addons/box/ edited to use new oauth system. Tests updated to reflect these changes.

Side Effects
==========
Tests obviated by changes removed.

Notes
=====
Before merge: 
* In the app settings on box.com, ensure that redirect_uri (in 'OAuth2 Parameters') ends in '/oauth/callback/box/'
* Make sure the script in #4395 is merged and run

### CODE REVIEW

File: website/addons/box/model.py [hunk 1]
```diff
@@ -1,28 +1,59 @@
...
+        client = BoxClient(CredentialsV2(
+            response['access_token'],
+            response['refresh_token'],
+            settings.BOX_KEY,
+            settings.BOX_SECRET,
+        ))
+
+        about = client.get_user_info()
+
+        url = 'https://app.box.com/profile/' + about['id']
```
Comment (lyndsysimon) [inline website/addons/box/model.py]: Not a big deal, but [new-style string formatting is preferred](http://cosdev.readthedocs.org/en/latest/style_guides/python.html#string-formatting).

File: website/addons/box/model.py [hunk 2]
```diff
@@ -340,114 +228,12 @@ def create_waterbutler_log(self, auth, action, metadata):
...
-                ).format(url=url)
-            #
-            return message
-
-    def after_delete(self, node, user):
-        self.deauthorize(Auth(user=user), add_log=True)
+            self.deauthorize(add_log=True)
         self.save()
+
+    def on_delete(self, node=None, user=None):
```
Comment (lyndsysimon) [inline website/addons/box/model.py]: Why the separate `on_delete` and `after_delete`?

If both are required, seeing as you're passing the same arguments, `on_delete = after_delete` should suffice.

File: website/addons/box/utils.py [hunk 3]
```diff
@@ -78,6 +86,40 @@ def log(self, action, extra=None, save=False):
             self.node.save()
 
 
+def refresh_oauth_key(external_account, force=False):
```
Comment (lyndsysimon) [inline website/addons/box/utils.py]: This doesn't look specific to Box - shouldn't it go somewhere in `website.oauth`?

Comment (mfraezz) [inline website/addons/box/utils.py]: `refresh_v2_token` comes from the `box.py` client library. Though it does need to be generalized, I agree. I think that work should go in a separate PR, as `website.oauth` doesn't currently have any token-refreshing capabilities and this PR is for Box.

File: website/addons/box/model.py (hunk 2 above)
Comment (mfraezz) [inline website/addons/box/model.py]: You're right. Both are required, but I wasn't sure what the correct thing to do here was, so I just picked one that worked.

File: website/addons/box/utils.py (hunk 3 above)
Comment (lyndsysimon) [inline website/addons/box/utils.py]: Fair enough. If you want to do that generalizing, let me know. Otherwise, if you could submit an Issue referencing it I'd appreciate it.

**Comment by lyndsysimon** [comment]
Second pass complete. Looks good to me, and the failing test doesn't seem relevant to this PR.

Paging @lbanner for disposition.

**Comment by lbanner** [comment]
Cool @lyndsysimon and @mfraezz ..... @sloria Let's fold this in when you are ready, sooner if that helps keep from code rot. Thanks.

**Comment by sloria** [comment]
@mfraezz Can you please get this up to date with develop? Hopefully that will fix the failing test.

File: website/addons/box/utils.py (hunk 3 above)
Comment (mfraezz) [inline website/addons/box/utils.py]: I'll do both

**Comment by mfraezz** [comment]
Yup, that fixed it.

**Comment by samanehsan** [comment]
Please delete the `website/addons/box/views/` directory since all of it's file have been removed.

**Comment by samanehsan** [comment]
Will we still need the script `refresh_box_tokens.py` after this refactor? It refreshes tokens on the BoxOauthSettings model which has been removed, so I'm not sure if this needs to be updated or if it can be deleted as well.

File: website/addons/box/static/node-cfg.js [hunk 4]
```diff
@@ -1,16 +1,7 @@
...
-        var name = item.data.path === 'All Files' ? '/ (Full Box)' : item.data.path.replace('All Files', '');
-        this.selected({
-            name: name,
-            path: item.data.path,
-            id: item.data.id
-        });
-        return false; // Prevent event propagation
-    }
-});
+require('./box.css')
```
Comment (samanehsan) [inline website/addons/box/static/node-cfg.js]: Missing a semicolon

**Comment by mfraezz** [comment]
`website/addons/box/views/` has been deleted, it just isn't automatically deleted when you check out a branch. See https://github.com/mfraezz/osf.io/tree/1b59e6eb84fd3c5bece20fa9423ca8a402fd0fef/website/addons/box

Good catch about the `refresh_box_tokens.py` script, it will need to be updated.

**Comment by samanehsan** [comment]
Ok, good to know. Don't forget to update the tests as well! They're in `scripts/tests/test_refresh_box_tokens.py`

**Comment by samanehsan** [comment]
Also, the redirect_uri listed in the box README.md is outdated. I believe it should be http://localhost:5000/oauth/callback/box/

File: website/addons/box/static/boxNodeConfig.js [hunk 5]
```diff
@@ -10,346 +10,142 @@ var $ = require('jquery');
...
-                'text-danger');
-            Raven.captureMessage('Could not GET Box settings', {
-                url: url,
-                textStatus: textStatus,
-                error: error
-            });
+var BoxNodeConfigViewModel = oop.extend(AddonNodeConfigViewModel, {
+    constructor: function(addonName, url, selector, folderPicker) {
+        var self = this;
+        self.super.super.constructor.call(self, addonName, url, selector, folderPicker);
```
Comment (samanehsan) [inline website/addons/box/static/boxNodeConfig.js]: Is there a reason you're calling `self.super.super.constructor` instead of `self.super.constructor` (the AddonNodeConfigViewModel constructor) here?

File: website/addons/box/static/boxUserConfig.js [hunk 6]
```diff
@@ -1,123 +0,0 @@
...
-                self.messageClass('text-info');
-            }, timeout);
-        }
-    };
-
-    /** Pop up confirm dialog for deleting user's access token. */
-    self.deleteKey = function() {
-        bootbox.confirm({
-            title: 'Disconnect Box Account?',
-            message: language.confirmDeauth,
```
Comment (samanehsan) [inline website/addons/box/static/boxUserConfig.js]: Please remove these auth messages since they are no longer being used: 
https://github.com/CenterForOpenScience/osf.io/blob/develop/website/static/js/osfLanguage.js#L62-#L68

File: website/addons/box/static/boxNodeConfig.js (hunk 5 above)
Comment (mfraezz) [inline website/addons/box/static/boxNodeConfig.js]: Yes, it's to circumvent the superclass' constructor (`AddonNodeConfigViewModel`) and go to its superclass (`FolderPickerViewModel). I remember that not doing so broke things, but I don't recall precisely what it broke, from 4 months ago. I'll look into it again

File: website/addons/box/static/boxUserConfig.js (hunk 6 above)
Comment (mfraezz) [inline website/addons/box/static/boxUserConfig.js]: This file was removed. The default user config file will be used, and it should have the correct language.

Comment (samanehsan) [inline website/addons/box/static/boxUserConfig.js]: Yeah, and since this file was the only place where those constants were being used, I think it's safe to delete them from `osfLanguage.js`

File: website/addons/box/serializer.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            'config': node.api_url_for('box_set_config'),
+            'emails': node.api_url_for('box_get_share_emails'),
+            'share': 'https://app.box.com/files/0/f/{0}'.format(self.node_settings.folder_id),
+            'deauthorize': node.api_url_for('box_remove_user_auth'),
+            'accounts': node.api_url_for('box_get_user_settings'),
+        }
+
+    @property
+    def addon_serialized_urls(self):
+        node = self.node_settings.owner
```
Comment (samanehsan) [inline website/addons/box/serializer.py]: Is this property necessary? It looks almost identical to `BoxSerializer.serialized_urls`.

File: website/addons/box/static/boxUserConfig.js (hunk 6 above)
Comment (mfraezz) [inline website/addons/box/static/boxUserConfig.js]: Gotcha. Misunderstood

File: website/addons/box/views.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
+"""Views fo the node settings page."""
```
Comment (samanehsan) [inline website/addons/box/views.py]: Small typo here -- this should say "for"

File: website/addons/box/utils.py [hunk 9]
```diff
@@ -78,6 +86,40 @@ def log(self, action, extra=None, save=False):
...
+        key = refresh_v2_token(settings.BOX_KEY, settings.BOX_SECRET, external_account.refresh_token)
+
+        external_account.oauth_key = key['access_token']
+        external_account.refresh_token = key['refresh_token']
+        external_account.expires_at = datetime.utcfromtimestamp(time.time() + key['expires_in'])
+        external_account.save()
+
+
+def get_folders(client):
+    """Gets a list of folders in a user's Box, including the root.
```
Comment (samanehsan) [inline website/addons/box/utils.py]: I'm not sure if this function is being used anywhere. Can it be removed?

File: website/addons/box/utils.py [hunk 10]
```diff
@@ -78,6 +86,40 @@ def log(self, action, extra=None, save=False):
...
+        'path': '',
+    }
+    folders = [root] + [
+        serializer.serialize_folder(each)
+        for each in metadata['contents'] if each['is_dir']
+    ]
+    return folders
+
+
 def handle_box_error(error, msg):
```
Comment (samanehsan) [inline website/addons/box/utils.py]: It looks like all usages of `handle_box_error` have been deleted -- can this function be removed as well?

File: website/addons/box/static/boxNodeConfig.js [hunk 11]
```diff
@@ -10,346 +11,143 @@ var $ = require('jquery');
...
+        var self = this;
+        //super.super used to prevent recursive requires w/ webpack
+        self.super.super.constructor.call(self, addonName, url, selector, folderPicker);
+        self.userAccountId = ko.observable('');
+        //externalAccounts
+        self.accounts = ko.observable([]);
+
+        self.messages.submitSettingsSuccess = ko.pureComputed(function() {
+            return 'Successfully linked "' + $osf.htmlEscape(self.folder().name) + '". Go to the <a href="' +
+                self.urls().files + '">Overview page</a> to view your files.';
```
Comment (samanehsan) [inline website/addons/box/static/boxNodeConfig.js]: This link should say "Files page".

File: website/addons/box/serializer.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+from box.client import BoxClient, BoxClientException
+
+
+class BoxSerializer(OAuthAddonSerializer):
+
+    def credentials_owner(self, user_settings=None):
+        return user_settings.owner or self.user_settings.owner
+
+    def serialize_folder(self, folder):
```
Comment (samanehsan) [inline website/addons/box/serializer.py]: Looks like this method is no longer being used -- can you delete this along with `test_utils/test_serialize_folder'?

File: scripts/refresh_box_tokens.py [hunk 13]
```diff
@@ -44,7 +46,7 @@ def main(delta, dry_run):
     try:
         days = int(sys.argv[2])
     except (IndexError, ValueError, TypeError):
-        days = 60 - 7  # refresh tokens are good for 60 days
+        days = 7  # refresh tokens that expire this week
```
Comment (samanehsan) [inline scripts/refresh_box_tokens.py]: Why was this changed?

**Comment by samanehsan** [comment]
Also, I found that when you click "Connect Account", then select "Deny Access to Box" instead of going through and connecting, you get a Value Error:  
screen shot 2015-09-17 at 10 49 05 am

**Comment by samanehsan** [comment]
Also, after connecting a Box account, when I go to connect a second account I am immediately taken to the "grant access to box" page instead of the login page (probably because I am still logged in to my box account). Would it be possible to always take the user to the login page when they click "Connect Account"? I think Mendeley does it this way.

**Comment by samanehsan** [comment]
Second pass complete @mfraezz! :octopus:

File: scripts/refresh_box_tokens.py (hunk 13 above)
Comment (mfraezz) [inline scripts/refresh_box_tokens.py]: It was previously set to refresh any tokens that expire at or before 53 days ago each time it was run. It's now set to refresh any tokens that expire at or before one week from the time it's run.

**Comment by mfraezz** [comment]
The `ValueError` is an issue with the underlying architecture. See https://github.com/CenterForOpenScience/osf.io/pull/4310. Updates will need to be made to the front-end after that's merged

**Comment by samanehsan** [comment]
Please change this error message (which shows up on both the user and project settings page) to use the `text-danger` class: 
screen shot 2015-10-05 at 9 24 59 am

File: website/addons/box/serializer.py (hunk 12 above)
Comment (samanehsan) [inline website/addons/box/serializer.py]: @mfraezz can you confirm whether or not this function is being used? I only see it used in it's test, so I think it can be removed.

**Comment by samanehsan** [comment]
Pass complete @mfraezz! :partly_sunny:

**Comment by samanehsan** [comment]
Thanks for making those changes @mfraezz! Done with CR1 :camel:

**Comment by mfraezz** [comment]
:+1:

File: website/addons/box/model.py [hunk 14]
```diff
@@ -208,7 +93,10 @@ def has_auth(self):
 
     @property
     def complete(self):
-        return self.has_auth and self.folder_id is not None
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
```
Comment (samanehsan) [inline website/addons/box/model.py]: Should this be `self.foreign_user_settings` instead of `self.user_settings`?

Comment (samanehsan) [inline website/addons/box/model.py]: Oh, nevermind...I just saw that it inherits a field named `user_settings`

File: website/addons/box/model.py [hunk 15]
```diff
@@ -1,194 +1,70 @@
...
-        return u'<BoxUserSettings(user={self.owner.username!r})>'.format(self=self)
 
+class BoxNodeSettings(StorageAddonBase, AddonOAuthNodeSettingsBase):
 
-class BoxNodeSettings(StorageAddonBase, AddonNodeSettingsBase):
+    oauth_provider = Box
+    serializer = BoxSerializer
 
-    user_settings = fields.ForeignField(
+    foreign_user_settings = fields.ForeignField(
```
Comment (samanehsan) [inline website/addons/box/model.py]: This field won't be needed after all of the migrations, right?

**Comment by chennan47** [comment]
@mfraezz Could you update this pr to solve the merge conflicts? Thanks.

File: website/addons/box/model.py [hunk 16]
```diff
@@ -303,94 +189,11 @@ def create_waterbutler_log(self, auth, action, metadata):
...
-                ).format(url=url)
-            #
-            return message
-
-    def after_delete(self, node, user):
-        self.deauthorize(Auth(user=user), add_log=True)
+            self.deauthorize(add_log=True)
         self.save()
+
+    on_delete = after_delete
```
Comment (chennan47) [inline website/addons/box/model.py]: on delete doesn't seems to be used anywhere.

Comment (mfraezz) [inline website/addons/box/model.py]: It's a super-super-superclass [override](https://github.com/mfraezz/osf.io/blob/feature/box-oauth-refactor/website/addons/base/__init__.py#L233-L235), used when [deleting](https://github.com/mfraezz/osf.io/blob/feature/box-oauth-refactor/website/addons/base/__init__.py#L209) nodes.

**Comment by chennan47** [comment]
CR2 done. Looks good.

File: scripts/box/connect_external_accounts.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+First run 
+    rm -rf website/addons/box/views/
```
Comment (sloria) [inline scripts/box/connect_external_accounts.py]: Why is this necessary?

File: scripts/box/connect_external_accounts.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+Then change the user_settings field of BoxNodeSettings to foreign_user_settings
```
Comment (sloria) [inline scripts/box/connect_external_accounts.py]: This is already done, correct?

Comment (sloria) [inline scripts/box/connect_external_accounts.py]: The docstring should explain the purpose and any pre-migration steps that need to be taken before the script is run. Take out anything that isn't necessary.

File: scripts/box/connect_external_accounts.py (hunk 17 above)
Comment (mfraezz) [inline scripts/box/connect_external_accounts.py]: `box/views` changed from being a folder with several files to a single `views.py`, but the folder and associated `.pyc` files will remain, causing the server to expect `BoxOAuthSettings` even though it's been removed.
